azurerm/internal/services/compute/virtual_machine_extension.go

```diff
@@ -0,0 +1,49 @@
+package compute
+
+import (
+	"fmt"
+
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+)
+
+type VirtualMachineExtensionID struct {
+	ResourceGroup  string
+	Name           string
+	VirtualMachine string
+}
+
+func ParseVirtualMachineExtensionID(input string) (*VirtualMachineExtensionID, error) {
+	id, err := azure.ParseAzureResourceID(input)
+	if err != nil {
+		return nil, fmt.Errorf("[ERROR] Unable to parse App Service ID %q: %+v", input, err)
+	}
+
+	virtualMachineExtension := VirtualMachineExtensionID{
+		ResourceGroup: id.ResourceGroup,
+	}
+
+	if virtualMachineExtension.Name, err = id.PopSegment("extensions"); err != nil {
+		return nil, err
+	}
+
+	if virtualMachineExtension.VirtualMachine, err = id.PopSegment("virtualMachines"); err != nil {
+		return nil, err
+	}
```

Comment:
since the ID's built up of multiple key-vaule pairs, the ordering could be important here in time; should we make this:

a) parse VM Name
b) parse name?